### PR TITLE
[FIX] pos_adyen, pos_restaurant_adyen: permissions

### PR DESCRIPTION
**Pull request by thvi-odoo**
- Adds some permissions to protect sensitive pos_adyen data and methods.
 - Reduces the risks of abuses of the pos_adyen features.

Those changes are not optimal. The whole system should be changed.

 - Fixes errors when requesting adjustment/capture to Adyen (in pos_restaurant_adyen).

task-id: 3500598

### CODE REVIEW

File: addons/pos_adyen/models/pos_payment_method.py [hunk 1]
```diff
@@ -6,23 +6,27 @@
...
 from werkzeug.exceptions import Forbidden
+from urllib.parse import parse_qs
 
 from odoo import fields, models, api, _
-from odoo.exceptions import ValidationError
+from odoo.exceptions import ValidationError, UserError, AccessError
 
 _logger = logging.getLogger(__name__)
 
+UNPREDICTABLE_ADYEN_DATA = 'UNPREDICTABLE'
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:18]: Thank you for the tip.
But for your second suggestion "the naming seems to imply that the data is random or unreliable", indeed it is unreliable.
By naming it "placeholder", we lose (in my opinion) the purpose of this constant: to define an Adyen field value that we cannot check.

File: addons/pos_adyen/models/pos_payment_method.py [hunk 2]
```diff
@@ -87,24 +73,137 @@ def get_latest_adyen_status(self, pos_config_name):
 
     def proxy_adyen_request(self, data, operation=False):
         ''' Necessary because Adyen's endpoints don't have CORS enabled '''
-        if data['SaleToPOIRequest']['MessageHeader']['MessageCategory'] == 'Payment': # Clear only if it is a payment request
+        self.ensure_one()
+        if not self.env.su and not self.user_has_groups('point_of_sale.group_pos_user'):
+            raise AccessError(_('You are not allowed to access to that.'))
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:78]: Do you mean "raise AccessError()" without any message ?

File: addons/pos_adyen/models/pos_payment_method.py [hunk 3]
```diff
@@ -87,24 +73,137 @@ def get_latest_adyen_status(self, pos_config_name):
...
+
+        is_payment_request_without_acquirer_data = operation == 'terminal_request' and self._is_valid_adyen_request_data(data, expected_payment_request_without_acquirer)
+        if not is_payment_request_without_acquirer_data and not is_payment_request_with_acquirer_data and not is_adjust_data and not is_cancel_data and not is_capture_data:
+            raise UserError(_('Invalid adyen request'))
+
         return self._proxy_adyen_request_direct(data, operation)
 
+    @api.model
+    def _is_valid_adyen_request_data(self, provided_data, expected_data):
+        if not isinstance(provided_data, dict) or set(provided_data.keys()) != set(expected_data.keys()):
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:171]: I personally prefer to explicitly write that we check the keys, but if you really want that change I can do it.

File: addons/pos_adyen/models/pos_payment_method.py [hunk 4]
```diff
@@ -87,24 +73,137 @@ def get_latest_adyen_status(self, pos_config_name):
...
+                        'AmountsReq': {
+                            'Currency': UNPREDICTABLE_ADYEN_DATA,
+                            'RequestedAmount': UNPREDICTABLE_ADYEN_DATA,
+                        },
+                    },
+                },
+            },
+        }
+
+        expected_payment_request_with_acquirer = copy.deepcopy(expected_payment_request_without_acquirer)
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:145]: Yes, that's what I thought at the beginning. But this whole method shouldn't exist, and I'm just patching it. I preferred not to create additional methods that would make this system as intended/justified/legitimate. By only putting the checks in proxy_adyen_request, the intention is to show that this logic is only for this (bad) method.
It's true that I still added some methods: _get_expected_message_header and _is_valid_adyen_request_data, but they don't have any impact and they were really needed to make the code readable. Here we just need the same dict with just 1 change.
But if you want it to be extracted, I can do it.

File: addons/pos_adyen/models/pos_payment_method.py [hunk 5]
```diff
@@ -6,23 +6,27 @@
 import random
 import requests
 import string
+import copy
 from werkzeug.exceptions import Forbidden
+from urllib.parse import parse_qs
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:11]: Just to be sure to understand, you are talking about the imports that I didn't write in the lexicographical order ? Is there any consequence or it's just a convention ?

File: addons/pos_adyen/models/pos_payment_method.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py]: Finally, the use of the "object()" for UNPREDICTABLE_ADYEN_DATA makes that the deepcopy doesn't work. So I extracted this in a method.

File: addons/pos_adyen/controllers/main.py [hunk 6]
```diff
@@ -27,6 +29,23 @@ def notification(self):
...
+                    pos_hmac = pos_hmac_metadata[0] if pos_hmac_metadata and len(pos_hmac_metadata) == 1 else None
+
+                    msg_header = data['SaleToPOIResponse']['MessageHeader']
+                    if not pos_hmac or not consteq(pos_hmac, payment_method._get_hmac(msg_header['SaleID'], msg_header['ServiceID'], msg_header['POIID'], data['SaleToPOIResponse']['PaymentResponse']['SaleData']['SaleTransactionID']['TransactionID'])):
+                        _logger.warning('Received invalid Adyen event notification (invalid hmac): \n%s', pprint.pformat(data))
+                        return
+
+                    pos_hmac_metadata_raw = 'metadata.pos_hmac='+pos_hmac
+                    safe_additional_response = adyen_additional_response.replace('&'+pos_hmac_metadata_raw, '').replace(pos_hmac_metadata_raw, '')
+                    data['SaleToPOIResponse']['PaymentResponse']['Response']['AdditionalResponse'] = safe_additional_response
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/controllers/main.py:45]: Thank you for the tip, and the reason why I hide the hmac is the same as explained in https://github.com/odoo/odoo/pull/140327#discussion_r1381585069

File: addons/pos_adyen/models/pos_payment_method.py [hunk 7]
```diff
@@ -87,24 +73,155 @@ def get_latest_adyen_status(self, pos_config_name):
...
+
+        if not is_payment_request_without_acquirer_data and not is_payment_request_with_acquirer_data and not is_adjust_data and not is_cancel_data and not is_capture_data:
+            raise UserError(_('Invalid Adyen request'))
+
+        if is_payment_request_with_acquirer_data or is_payment_request_without_acquirer_data:
+            acquirer_data = data['SaleToPOIRequest']['PaymentRequest']['SaleData'].get('SaleToAcquirerData')
+            msg_header = data['SaleToPOIRequest']['MessageHeader']
+            metadata = 'metadata.pos_hmac=' + self._get_hmac(msg_header['SaleID'], msg_header['ServiceID'], msg_header['POIID'], data['SaleToPOIRequest']['PaymentRequest']['SaleData']['SaleTransactionID']['TransactionID'])
+
+            data['SaleToPOIRequest']['PaymentRequest']['SaleData']['SaleToAcquirerData'] = acquirer_data + '&' + metadata if acquirer_data else metadata
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py:150]: For the same reason as explained in https://github.com/odoo/odoo/pull/135499#discussion_r1408963484, I would prefer to keep my approach.

File: addons/pos_adyen/models/pos_payment_method.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/models/pos_payment_method.py]: Indeed I totally agree, I wanted to use AccessDenied but its description dissuaded me. I will therefore use it as you suggested, thank you.

File: addons/pos_adyen/controllers/main.py [hunk 8]
```diff
@@ -2,14 +2,16 @@
...
 from odoo import fields, http
 from odoo.http import request
+from odoo.tools import consteq
 
 _logger = logging.getLogger(__name__)
 
 
 class PosAdyenController(http.Controller):
-    @http.route('/pos_adyen/notification', type='json', methods=['POST'], auth='none', csrf=False)
+    @http.route('/pos_adyen/notification', type='json', methods=['POST'], auth='public', csrf=False)
```
[the remark this answers is not in the document]
Comment (thvi-odoo) [reply addons/pos_adyen/controllers/main.py:14]: Ok thank you

**Review by caburj — APPROVED** [review]
@thvi-odoo Thank you for this PR and good job!

@odony I appreciate the good guidance to my colleague. I did a quick check and I'm okay with the changes so we're merging.

robodoo r+

fw-bot up to saas-16.4

**Comment by caburj** [comment]
robodoo r+

**Comment by caburj** [comment]
robodoo r+

**Pull request closed without merging**